x/poolmanager/router.go

```diff
@@ -273,67 +273,71 @@ func (k Keeper) MultihopEstimateOutGivenExactAmountIn(
 	return tokenOutAmount, err
 }
 
-// MultihopSwapExactAmountOut defines the output denom and output amount for the last pool.
-// Calculation starts by providing the tokenOutAmount of the final pool to calculate the required tokenInAmount
-// the calculated tokenInAmount is used as defined tokenOutAmount of the previous pool, calculating in reverse order of the swap
-// Transaction succeeds if the calculated tokenInAmount of the first pool is less than the defined tokenInMaxAmount defined.
+// RouteExactAmountIn processes a swap along the given route using the swap function corresponding
+// to poolID's pool type. This function is responsible for computing the optimal input amount
+// for a given output amount when swapping tokens, taking into account the current price of the
+// tokens in the pool and any slippage.
+// Transaction succeeds if final amount out is greater than tokenOutMinAmount defined
+// and no errors are encountered along the way.
 func (k Keeper) RouteExactAmountOut(ctx sdk.Context,
 	sender sdk.AccAddress,
 	route []types.SwapAmountOutRoute,
 	tokenInMaxAmount sdk.Int,
 	tokenOut sdk.Coin,
 ) (tokenInAmount sdk.Int, err error) {
 	isMultiHopRouted, routeSwapFee, sumOfSwapFees := false, sdk.Dec{}, sdk.Dec{}
+	// Ensure that provided route is not empty and has valid denom format.
 	routeStep := types.SwapAmountOutRoutes(route)
 	if err := routeStep.Validate(); err != nil {
 		return sdk.Int{}, err
 	}
 
+	// ? should we have defer in RouteExactAmountIn too?
```

Comment:
I do not know the answer for this :(  Once you find out, would you be able to add the reason we defer to the comment as well?

Reply:
Yes, I believe this should also happen in RouteExactAmountIn as well (we should also add panic recovering tests like what was done here https://github.com/osmosis-labs/osmosis/pull/4248)

Reply:
@mattverse we discussed this in our CL call today, and we removed the defer from `RouteExactAmountIn` for protorev, we came to a conclusion that we should have it here, unless it affects any other part of the code